lib/vrl/stdlib/src/parse_host.rs

```diff
@@ -0,0 +1,211 @@
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use tldextract::*;
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct ParseHost;
+
+impl Function for ParseHost {
+    fn identifier(&self) -> &'static str {
+        "parse_host"
+    }
+
+    fn summary(&self) -> &'static str {
+        "parse a URL or domain name string"
+    }
+
+    fn usage(&self) -> &'static str {
+        indoc! {r#"
+            Parses the provided `value` into TLD, registered domain, and subdomains.
+
+            By default includes private domains.
+        "#}
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::BYTES,
+                required: true,
+            },
+            Parameter {
+                keyword: "private_domains",
+                kind: kind::BOOLEAN,
+                required: false,
+            },
+            Parameter {
+                keyword: "update_cache",
+                kind: kind::BOOLEAN,
+                required: false,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "parse domain name",
+                source: r#"parse_host!("www.example.com")"#,
+                result: Ok(indoc! {r#"
+                    {
+                        "domain": "example",
+                        "subdomain": "www",
+                        "suffix": "com"
+                    }
+                "#}),
+            },
+            Example {
+                title: "parse URL",
+                source: r#"parse_host!("https://www.example.com/s?q=foobar")"#,
+                result: Ok(indoc! {r#"
+                    {
+                        "domain": "example",
+                        "subdomain": "www",
+                        "suffix": "com"
+                    }
+                "#}),
+            },
+            Example {
+                title: "ignore private domain",
+                source: r#"parse_host!("internal.us-east-1.elb.amazonaws.com", private_domains: false)"#,
+                result: Ok(indoc! {r#"
+                    {
+                        "domain": "amazonaws",
+                        "subdomain": "internal.us-east-1.elb",
+                        "suffix": "com"
+                    }
+                "#}),
+            },
+        ]
+    }
+
+    fn compile(&self, state: &state::Compiler, mut arguments: ArgumentList) -> Compiled {
+        let data_dir: String = state
+            .get_external_context::<std::path::PathBuf>()
```

Comment:
It would be worth creating a wrapper for the path:

```
struct DataDir(PathBuf);
```

so then if another function needs a path that means something else it won't collide with this.

```suggestion
            .get_external_context::<DataDir>()
```

Reply:
Indeed - I wasn't super sure where I wanted to create that, and wanted to make sure this direction was approved before sinking some more work into it 😄